### ISSUE TITLE
[Discussion] Better Way to Declare Globals in JavaScript files

### ISSUE THREAD

**Issue by mjbvz** [Microsoft/TypeScript#15626]
**Problem**
Using the new `checkJs` feature, it is currently difficult to tell TypeScript about a global variable:

```js
// @ts-check

myGlobal

window.myVar = 'foo';
```

<img width="458" alt="screen shot 2017-05-05 at 4 44 20 pm" src="https://cloud.githubusercontent.com/assets/12821956/25767871/268f4b00-31b2-11e7-988f-fe9f14165c60.png">

The only way around this that I know is to create a `.d.ts` file that defines these:

```ts
interface Window {
    myVar: string;
}

declare var myGlobal: string;
```

JavaScript users should not have to write `*.d.ts` files.

**Possible Approaches**
@egamma proposes that may be able to use the `global` comment from ESlint http://eslint.org/docs/user-guide/configuring : `/* global var1, var2 */`

I'd also be interested to see if we could put type information in a jsdoc somehow, something like:

```js
/**
 * @global {string} myGlobal 
 */
```

**Comment by mhegazy** [comment]
> JavaScript users should not have to write *.d.ts files.

Why is that? Js users are already using dts files from @types; so does not seem totally foreign concept.

**Comment by Zarel** [comment]
> Why is that?

I mean, it's a long time since the C/C++ days when we had to manually write header files. There's no reason to force a programmer to write a file describing a variable that TypeScript's already aware of and understands the typing of.

The job of a computer is to do easily-automated tasks for humans. If you are telling a human to do an easily-automated task for a computer, something is seriously wrong.

**Comment by mhegazy** [comment]
The OP suggests using jsdoc. So some one has to write something. Why not in a .d.ts file

**Comment by DanielRosenwasser** [comment]
Yes, you can use `.d.ts` files, but the point of this feature is lower barrier to entry. Plus, if I'm writing a quick one-off script, it can be nice to not worry about such details.

**Comment by weswigham** [comment]
If you don't want to write a `.d.ts`, there's no need to overthink it, you can just annotate a var declaration in the global scope, like so:
```js
/**
*  @type {string}
*/
var myString;
```
then later statements like: 
```js
console.log(myString.toUpperCase());
```
will be typechecked.

Even if you're writing modules, you can add a file containing globals like this to your compilation - just don't write any imports or exports and it should be global scoped as expected.

Unfortunately, as far as I'm aware there's no way to patch definitions onto existing types at present using only jsdoc syntax, so for that you'd have to use a `.d.ts`. - the correct way, I believe, would be to interpret `@memberof` jsdoc keywords as indicating a type patching operation, and then using those, but AFAIK TS doesn't treat `@memberof` in any special way right now.

**Comment by egamma** [comment]
When discussing the options it would also be good to provide a quick fix for such an error. 

(we did that in VS Code before we switched to Salsa by providing a quick fix that added eslint global comment /* global var */ to the file)

**Comment by mjbvz** [comment]
I've added some documentation to VSCode on the `d.ts` workaround for global variables: https://github.com/Microsoft/vscode-docs/pull/991 

My main concern with `d.ts` files is that regular JavaScript programmers may not have any experience with them. JSDocs or eslint style directives may be less intimidating and they also wouldn't require creating a `jsconfig.json` like the `d.ts` based approach does

**Comment by Zarel** [comment]
@weswigham 

> Even if you're writing modules, you can add a file containing globals like this to your compilation - just don't write any imports or exports and it should be global scoped as expected.

Unfortunately, this doesn't work for `@typedef` types/interfaces.

I'm still not sure how I use types defined in other files at all...

**Comment by rbuckton** [comment]
[usejsdoc.org](http://usejsdoc.org) specifies two existing JSDoc tags that could be useful here:
* [`@global`](http://usejsdoc.org/tags-global.html) - does not support name/type itself
* [`@var`](http://usejsdoc.org/tags-member.html#examples) - example: `/** @var {*} myGlobal */`

**Comment by mhegazy** [comment]
based on the discussion in https://github.com/Microsoft/TypeScript/issues/15747, we need to support `@var` as the JSDoc representation of `declare var ..` in the global scope.

**Comment by mhegazy** [comment]
adding globals to window is tracked under #12902

**Comment by thw0rted** [comment]
As a vanilla-JS user who just started using `checkJs`, it would be really nice if my eslint comment (`/* global myVar */`) worked with TS so I didn't have to document my global in two different ways for two different checkers.  I hope that the TypeScript team supports the goal of `checkJs` working on any valid JS without having to add a `.d.ts` file to your build just to make the checker happy.

**Comment by megakoresh** [comment]
In my opinion globals shouldn't be so complicated. Why not simply declare them in `tsconfig.json` like so:
```
...
"globalModules": [
  "./global.js"
]
...
```
where `global.js` could be
```
module.exports.global1 = 'my amazing global string';
//or if using ES6 modules
const global1 = 'my amazing global string';
export global1;
```
It would be simple enough for both TS and JS users, would be consistent with how VSCode is set up for JS projects, and is flexible enough. After all with TypeScript some kind of module system is always assumed, so why not use it?

**Comment by megakoresh** [comment]
@mhegazy Can we resume this discussion please? I think it is becoming more relevant with the rising popularity of vscode.

**Comment by mhegazy** [comment]
The issue is still open, it is still on our list of feature requests.

**Comment by megakoresh** [comment]
Yes, I understand, but if it appears as if this is still under discussion and the look of final implementation is still not decided. I wanted to see if there is some consensus made on how the final feature would look.

**Comment by mhegazy** [comment]
>  but it appears as if this is still under discussion and the look of final implementation is still not decided.

https://github.com/Microsoft/TypeScript/issues/15626#issuecomment-300633744

**Comment by greg-ashby** [comment]
+100 for the suggestions to use the eslint `/* global ... */` option. We use that already for eslint so it would be great to not have to declare it second way.

**Comment by mohsen1** [comment]
Please support JSDoc for this. having a `.d.ts` file works but you either have to put all of globals in one place or make a `.d.ts` file per JS file which is annoying. Having global declaration near where global is being used makes code cleaner as well. It will be similar to how ESLint deals with this 

Example code that uses a ESLint comment to declare globals to surpass ESLint error:

https://github.com/webpack/webpack/blob/fde018300aa52262c384e937c408d5dd97d62951/lib/webworker/WebWorkerMainTemplate.runtime.js#L5

**Comment by TotallyInformation** [comment]
Any movement on this proposal?

**Comment by xileftenurb** [comment]
About this, I did not find any ways to add a property to the node global object without modifing the `NodeJS.global` definition who is common for every project on a computer (and not in the project.)

the node `global` var is already declared, so I cannot use "@type" anotation as declaration
I cannot use a d.ts since it already existe, outside the scope of the project.

anyones as an idea on how to fix that?

I would like to be able to document code as simple as
`global.__rootdir = __dirname`
in node so i can use `global.__rootdir` in any file of my project.

**Comment by dcsan** [comment]
@xileftenurb I have exact same problem with some code I inherited. Did you find a workaround?

**Comment by xileftenurb** [comment]
the only workaround I found is to do at the start of the file who use the global :
```
// @ts-ignore
const rootdir = global.__rootdir
```
and to use the newly created variable in the file

**Comment by thw0rted** [comment]
If you look upthread, they discuss the correct syntax for declaring globals in a `.d.ts` (ambient type definition) file.  You'd then configure your `tsconfig.json` to point to that file, and subsequent runs of `tsc` will have type information for the globals you need.  You say you can't use a `.d.ts` file but I don't understand why not?

**Comment by dcsan** [comment]
> You'd then configure your tsconfig.json to point to that file

can you explain how to do this?
I did try editing a `.d.ts` but it didn't seem to get picked up.
[reference](https://www.typescriptlang.org/docs/handbook/tsconfig-json.html)

also partially the problem for me is that someone is using `global.X = {};` in a project I'm working on, and so i need to declare that `global` to have an `X` which is a node thing usually only declared in one place on your system.

**Comment by thw0rted** [comment]
This seems to work for me:

````
/* tsconfig.json */
{
  "compilerOptions": {
    ....
    "allowJs": true,                       /* Allow javascript files to be compiled. */
    "checkJs": true,                       /* Report errors in .js files. */
    ....
    "typeRoots": ["./typings", "./node_modules/@types"],                       /* List of folders to include type definitions from. */
    ....
  }
}

/* typings/globals.d.ts */
declare namespace NodeJS {
    interface Global {
        x: number;
    }
}

/* src/main.js */
console.log(global.x); // no error
console.log(global.y); // Property 'y' does not exist on type 'Global'
````

The thing to be aware of is that the typings for NodeJS (`@types/node`) that get included by default have the following:

````
declare var global: NodeJS.Global;
....
declare namespace NodeJS {
  interface Global {
    ....
    process: Process;
    root: Root;
    // etc etc
  }
}
````

This means it's declaring that `global` is a global variable, of type `NodeJS.Global`, and it defines the shape of that interface.  When you make another `.d.ts` file with its own `declare namespace NodeJS { interface Global { .... } }`, you're *augmenting* that interface with additional properties.  Both of these declarations apply to every file the compiler touches -- the types are merged.  This means that whoever provides `global.X = {}` can also include a `.d.ts` file that adds `X` to the `NodeJS.Global` interface.  The mechanism for specifying where that `.d.ts` file goes varies depending on how you're importing the library.

**Comment by dcsan** [comment]
thanks! that seems to be working!

**Comment by TotallyInformation** [comment]
Apologies that I'm not familiar enough as yet with Typescript but I have a couple of folders in my Node.js app that are actually front-end code and so have radically different globals to the rest of the code.

What would be the correct way to make sure I can check them properly? I am reluctant to put a tsconfig.json file in their root folder as those folders are actually made outwardly available via ExpressJS - though I can do so if needed.

**Comment by thw0rted** [comment]
@TotallyInformation you need to make separate projects for frontend vs backend, with their own `tsconfig.json`.  You can use the `typeRoots` option (as seen in my previous comment here) to keep the `.d.ts` files in a directory that's different from your JS sources.

Further "how do I" questions would probably be better for StackOverflow -- nothing personal, but we're drifting from the original intent of the issue, which is that (IMHO) TypeScript should provide an inline syntax for declaring globals that doesn't require introduction of a separate `.d.ts` file.  I'd like to see TS support the comment syntax used by eslint.

**Comment by TotallyInformation** [comment]
Thanks for the steer @thw0rted. No intent to turn this into a QA session :)

**Comment by dcsan** [comment]
@TotallyInformation why not? this is very helpful and this issue ranks high in google for the question. mute the thread if you prefer, it is a [Discussion]

**Comment by thw0rted** [comment]
Discussion means that this is the place to debate how best to implement a new or improved feature in Typescript.  It does not mean that the thread is open for questions about how to use existing capabilities.  If you click the "New Issue" button, you'll see a prompt that includes the text

> The issue tracker is not for questions. Please use Stack Overflow or other resources for help writing TypeScript code.

The presence of a "discussion" tag here does not change that guidance.

**Comment by heroboy** [comment]
How I can export a class in `window`?
For example:
```
(function(){
    class Foo()
    {
    }
   window.Foo = Foo;
})
```
If I use `.d.ts`, I must declare the `class Foo` again in `.d.ts`.

**Comment by Paril** [comment]
I'm hoping this still gets some traction. It'd seem like a fairly easy thing to have JS files in your workspace with @global's apply globally like they should; maybe I'm missing some detail here, but VSCode already picks up global functions and classes from various parts of my workspace. Being able to just plop in a @global on a definition and expect it to work as if it's part of globalThis would be amazing/preferred.

**Comment by kinsi55** [comment]
@thw0rted I have a global variable which is set to a class that is documented trough JSDoc, do you know if / how it would be possible to "backlink" from the `.d.ts` file to something thats defined in source?

My main issue is that I cannot use the globally defined class as a type in source files - JSDoc perfectly recognizes it when I generate docs using it, but TS / VSC has no idea what it is, eventho the global variable is `Foo.Bar` and the class, which is called Bar, is a `@memberof Foo`. I need to explicitly add a `@type {import('../sourcefile')} Foo.Bar` for it to be picked up by VSC, and even then, that does not resolve the issue of the global variable being unknown / having the incorrect type.

**Comment by bassem-mf** [comment]
The code that @heroboy posted above is very common in JavaScript programs. A lot of people wrap all the code in the JS file in a self executing function.
Here is what I do currently to document my global functions defined inside other functions.
```
/**
* @callback MyNamespace1.MyNamespace2.MyNamespace3.myFunction
* @param {string[]} arg1
* @param {any} arg2
* @param {number} arg3
*/

(function () {
    'use strict';

    /** @type MyNamespace1.MyNamespace2.MyNamespace3.myFunction */
    window.MyNamespace1.MyNamespace2.MyNamespace3.myFunction =
        function (arg1, arg2, arg3) {
            
        };

    // Other code that should not polute the global namespace.
})();

/** @type MyNamespace1.MyNamespace2.MyNamespace3.myFunction */
window.MyNamespace1.MyNamespace2.MyNamespace3.myFunction;
```
You see there is a lot of repetition and the code is not very elegant. But this is the best way I could find to get intellisense in Visual Studio. I hope we will have a better solution in the future based on `@global` or `@var`.

**Comment by olleharstedt** [comment]
NB: The Flow syntax to do this is (using jQuery as an example):

    /*:: declare var $: Object */

This can be done inline in any .js file using `// @flow`.

**Comment by tnguyen14** [comment]
> The only way around this that I know is to create a .d.ts file that defines these

I tried this but it's still not working. Does anyone know how I can get `.d.ts` file to work?

I have tried both of these in a `global.d.ts` file

```ts
declare var myGlobal: string;
```

```ts
interface Window {
  myGlobal: string;
}
```

**Comment by bfulop** [comment]
Hi @tnguyen14 ,

I had a similar issue as you. In my experience if your ambient type declaration file contains an `import` or `export` statement, then it's no longer considered to be a global declaration file.

**Comment by dmchurch** [comment]
I sure do hate necroposting an issue like this, but from what I can tell, it _still_ isn't resolved despite the [seven-year-old resolution on using the @var syntax](https://github.com/microsoft/TypeScript/issues/15626#issuecomment-300633744). Why is there still no way to declare global variables using JSDoc? We know the JSDoc syntax, we know the equivalent TS syntax. Why have these not been linked together yet?

For context, not that it particularly matters given that both the problem and the solution are fully specified, I'm trying to make a toy Electron project, using their `contextBridge` functionality to expose new globals to the renderer's JS environment. I do not have, nor do I want, a jsconfig.json file, a .d.ts file, a VSCode extension/configuration, or a package.json setting. I'm very familiar with all those workarounds. I am not looking for a _workaround_. All that I want is to have some way to tell the editor "in this file, the identifier `foo` is available in the global namespace, with type `bar`". This is the same issue I run into when I try to add typings to a one-off file intended to run outside the typical environment, like Worker or ServiceWorker code in a project that mostly runs in a DOM Window environment, and it frustrates me just as much then.

If it's simply a matter of "no one has done the work yet" I'd be delighted to submit a PR for it, but my experience with corporate-owned software projects like this one is that it is 100% futile to submit a PR without someone on the other side ready and waiting to receive it. That plus, the fact that this would be such a simple issue to knock out _and yet it's been seven years_ tells me that the issue isn't getting the work done, the issue is getting it accepted.

So? What's the hold-up? Is there anything that can be done to move this along, or is this just a pocket veto by the Typescript team?

**Comment by trusktr** [comment]
This would be so useful.

When writing Custom Elements with JS + JSDoc, I have one file per custom elements, for example for two elements `<my-element>` and `<other-element>` I have these two files:

```js
my-element.js
other-element.js
```

where one has a `MyElement` class and the other has a `OtherElement` class, one class per file per element definition.

To define type definitions for these elements so that they are visible in DOM APIs, for example so that we get this:

```js
const el = document.querySelector('my-element') // type of `el` is `MyElement`

/** @type {MyElement} */
const el2 = el // ok, el is assignable

/** @type {OtherElement} */
const el3 = el // type error, el is not OtherElement but MyElement
```

we would need to write definitions like this if we were using TypeScript files:

```ts
// my-element.js
class MyElement extends HTMLElement {...}
customElements.define('my-element', MyElement)

declare global {
  interface HTMLElementTagNameMap {
    'my-element': MyElement
  }
}
```
```ts
// other-element.js
class OtherElement extends HTMLElement {...}
customElements.define('other-element', OtherElement)

declare global {
  interface HTMLElementTagNameMap {
    'other-element': OtherElement
  }
}
```

But as you know, we can't do this in a JS file using JSDoc. So, what we have to do is *double* the number of files that we have, so that we can place these definitions in `.d.ts` files:

```js
my-element.js
my-element-types.d.ts
other-element.js
other-element-types.d.ts
```

(Note, the names must not be the same, for example `my-element.d.ts` would cause TS to see the file as a declaration for the sibling file, hence I have added `-types` to prevent this from happening, and instead the files are seen as files containing global types).

If we have 50 elements, now we have to double our files to 100, cluttering up the workspace. It would be great not to have to double the amount of files.

An alternative is to stick element definition all in a single file, but then that becomes less portable, for example you can't simply copy/paste files from project to project, but you have to specifically extract pieces of the .d.ts file that has the element defs.

Note, in related issues such as 

- https://github.com/tc39/proposal-type-annotations/issues/176
- https://github.com/tc39/proposal-type-annotations/issues/202
- https://github.com/microsoft/TypeScript/issues/9694
- and others

having a more useful comment space for writing TS types could help. For example, picking one of those syntax proposals at random, we could have something like the following in a JS file:

```ts
// other-element.js
class OtherElement extends HTMLElement {...}
customElements.define('other-element', OtherElement)

/*::
declare global {
  interface HTMLElementTagNameMap {
    'other-element': OtherElement
  }
}
*/
```

**Comment by LoganTann** [comment]
Also, .d.ts files might not be supported by vscode in some edge cases (having files without the js extension) : https://github.com/microsoft/vscode/issues/37772

The only way to fix this would be to have this proposal for jsdoc

**Comment by archive-20260622** [comment]
My team's environment doesn't support .d.ts files. For a purely JSDoc solution, I found that this works:

```js
// Include this file in rootTypings in tsconfig.json.

/** @type {import('./SomeGlobalDefinition.js')} */
var SomeGlobalDefinition;

// ...more of these definitions for all the globals in the project...
```

I also wrote a bash script to auto-generate this file by inspecting a directory of global definitions, since in my team's environment, the global names are the same as the file names.

**Comment by Arlen22** [comment]
Lol, ChatGPT suggested this to me today as an alternative to JSDoc, and linked me here. I sure wish it was a thing!

```
/*~ <-- notice the tilde
 * declare global {
 *   interface Something {
 *     property: string;
 *     property2: string;
 *   }
 * }
 */
```

**Comment by airtonix** [comment]
it'd be amazeballs if this was at least a thing: 

```js
/**
 * @type {import('https://esm.sh/react@19.1.0').React} React
 */
```

Even better would be 

```js
/**
 * @type global {import('https://esm.sh/react@19.1.0').React} React
 */
```

Why? 

- don't want to install typescript in the project
- don't want to install nodejs
- don't want to use npm (or any js pkg manager)

why? 

> reasons.
